### PR TITLE
Implement fixtures testing for tx-indexer

### PR DESCRIPTION
**Pull request by szg251**
- updated to plutus-ledger-api v2 (with Conway types, csl conversions, Display instances, etc)
- separated tx-indexer-testsuite into it's own module
- implemented fixtures testing (writing and reading events from files)

### CODE REVIEW

File: extras/tx-indexer-testsuite/process-compose.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+processes:
+  tests:
+    command: cargo test
+    availability:
+      exit_on_end: true
+    depends_on:
+      db: 
+        condition: process_healthy
+
+  db:
```
Comment (chfanghr) [inline extras/tx-indexer-testsuite/process-compose.yaml:12]: Huh, we do not handle the initialization of the database here?

File: tx-indexer/src/indexer.rs [hunk 2]
```diff
@@ -1,87 +1,210 @@
...
+async fn source_from_files(
+    handler: impl EventHandler,
+    dir_path: PathBuf,
+) -> Result<TxIndexer, anyhow::Error> {
+    use futures::stream::{StreamExt, TryStreamExt};
+    use tokio::fs;
+    use tokio::runtime::Runtime;
+    use tokio_stream::wrappers::ReadDirStream;
+
+    let file_stream = ReadDirStream::new(fs::read_dir(dir_path).await.map_err(|err| anyhow!(err))?);
```
Comment (chfanghr) [inline tx-indexer/src/indexer.rs:175]: How do you make sure that the events are replayed in order? I don't think ``ReadDir` ensures the order of entries it spits out.

**Review by chfanghr — APPROVED** [review]
Thanks for fixing all the imports haha

**Comment by chfanghr** [comment]
Currently it looks like the fixture tests completely bypass the ours stack right? In e2e tests we can probably check the database to make sure that the events are correctly recorded by the indexer, right after we generate said events.

File: extras/tx-indexer-testsuite/build.nix [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+          devShellHook = config.settings.shell.hook + ''
+            ln -sf ${../../tx-indexer}/lib-migrations
+          '';
+
+          buildInputs = [ pkgs.postgresql_16.lib ];
+
+          devShellTools = commands.devShellTools ++ [
+            pkgs.diesel-cli
+            pkgs.postgresql_16
+            pkgs.process-compose
```
[the remark this answers is not in the document]
Comment (szg251) [reply extras/tx-indexer-testsuite/build.nix:43]: oh, nice. This could be the way to run tests on CI

[the remark this answers is not in the document]
Comment (szg251) [reply extras/tx-indexer-testsuite/build.nix:46]: look at this: https://community.flake.parts/services-flake/postgresql

File: extras/tx-indexer-testsuite/process-compose.yaml (hunk 1 above)
Comment (szg251) [reply extras/tx-indexer-testsuite/process-compose.yaml]: now we do

File: tx-indexer/src/indexer.rs (hunk 2 above)
Comment (szg251) [reply tx-indexer/src/indexer.rs]: fixed

File: extras/tx-indexer-testsuite/build.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (szg251) [reply extras/tx-indexer-testsuite/build.nix]: I couldn't figure out how to run process-compose on the CI without reinventing the whole world, so I left it hanging, maybe another day...

[the remark this answers is not in the document]
Comment (chfanghr) [reply extras/tx-indexer-testsuite/build.nix]: I will look into it when I have time. Doesn't look so hard, we just need to wrap the process compose stack in a systemd service and run the test in another one shot service, and run everything in a nixos test.

[the remark this answers is not in the document]
Comment (szg251) [reply extras/tx-indexer-testsuite/build.nix]: What do you think which is simpler (more maintainable), wrapping the process-compose, or using native NixOS e2e testing, as we did elsewhere?

**Pull request merged**